### PR TITLE
Add neutral colormap and nogrey to palette tonemap

### PR DESCRIPTION
**Pull request by chetgurevitch**
I brought @PixelEater2018's PR (#720) up to what should hopefully be mergeable condition after discovering it fixed some tonemap issues I was seeing in doom2 map2. The code is unchanged outside of moving the language.enu changes to language.csv.

### CODE REVIEW

**Comment by coelckers** [comment]
I'll leave the review to dpJudas. However, regarding the texts, please read the relevant post on the forum. If you add text to language.csv it will be lost the next time that file gets updated from its master on Google Docs. If you need to add strings to the menu, please post the needed content as text here in a comment on the PR, so it can be moved to the master table.

**Comment by drfrag666** [comment]
If the new tonemap formula is right at least i could port that to my builds.

**Comment by dpjudas** [comment]
If you mean the change to the shader, then I'm not sure that is correct.

Most of it is just a different way of typing it (shifts vs multiplication). It may vary between hardware which is actually fastest (older GPU hardware might prefer the multiply). The removal of +0.5 looks wrong though. It was there to round the number and can't just be added to the multiplication.

**Comment by chetgurevitch** [comment]
The following strings are needed for the menu

GLPREFMNU_PALTONECMAP = "Tonemap Palette Colormap";
GLPREFMNU_PALTONENOGREY = "Prevent Grey Shadows";

**Comment by chetgurevitch** [comment]
@dpjudas you were right about the 0.5 shift, fixed that and the results now match up more closely to softpoly than in upstream or the previous version. Made some minor formatting tweaks as well.

**Comment by PixelEater2018** [comment]
Definitely double check this, but if you force it to round up with the +0.5 it causes the lowest index number to be 4161. If correct that means black will be fetched from the LUT at coordinates 65,8 instead of 0,0. Multiplying by 63.5 instead was the only way I found to fix SKY2 as 63 & 64 just moved the glitchiness to other colour ranges.

**Comment by dpjudas** [comment]
ivec3 c = ivec3(clamp(color.rgb, vec3(0.0), vec3(1.0)) * 63.0 + 0.5);

63.5 cast to int becomes 63 (rounds towards zero). Thus the integer range is [0-63]. The purpose of adding the 0.5 is to make it do normal rounding, which will work for all positive numbers.

We multiply with 63 because the value for 1.0 is 63.

It is possible the software renderer does not round correctly (cast/shift without adding the 0.5 first). In the past I've solved several such precision bugs that resulted into ugly stuff like the dancing sprites bug. If this is the case it really should be fixed in the software renderer and not the shader as proper rounding improves visuals.

Note that to fully match the software renderer converting the color in itself is not enough. The light level needs to be banded as well (Rachael did one attempt of this). And then the LUT itself also lacks the precision. Rachael did do some attempts at 'calibrating' the LUT to get closer, but ultimately it will always be an approximation as 6 bits just isn't enough. You need a 8 bit LUT for it to fully work (64 MB texture).

**Comment by PixelEater2018** [comment]
Seems I've had integer casting a little backwards then. I'd been thinking that if the fractional was .5 then the number would round _upward_! Thanks for the revelation dpJudas :)

**Comment by chetgurevitch** [comment]
Rebased onto the latest master. This should be ready to go, though ironically it doesn't fix either of the issues that prompted it.

Using 63 + 0.5 (the current pr) I can reproduce the issue @PixelEater2018 was having with sky2 over both opengl and vulkan, but only on nvidia, specifically the proprietary linux driver. I haven't tested this with either nouveau or nvidia's windows drivers yet, but the issue doesn't occur with intel's mesa drivers or amd's windows drivers.
sky2-banding

There's also still excessive blue and green banding on the ground at spawn in doom map02, which is what originally lead me to this PR. Here's a comparison between obverse palette, reverse palette, and software in that order.
map02-banding-obverse
map02-banding-reverse
map02-banding-software

**Comment by coelckers** [comment]
Since I do not use the tonemap feature I do not think I am the right person to sign off on this. I added the strings to the spreadsheets, so if this doesn't get an ok, I'll delete them later.

**Comment by madame-rachelle** [comment]
I tested this. I have always been a bit hesitant to green-light this, to be quite honest. But in-game visually, the new changes don't look so bad.

My biggest concern is with the lack of option to switch to the old algorithm. This might be one case where it's not needed - I can't be the best judge of that, as I only use this rarely, it's not my primary way of playing GZDoom, so I am really out of touch with anyone who would actually use this.

So yeah, my sign-off on this is ... "kinda? maybe?..." Not a very certain one here.

**Comment by chetgurevitch** [comment]
With the 63.5 change reverted, the doom colormap shouldn't actually have changed whatsoever. I diffed screenshots of spawn at e1m1 from this branch and from the latest release with imagemagick, and they were identical outside of the eyes on the armor pickups and the acid in the barrel, which were just on different frames of their animations.

Really all this does now is add the neutral colormap and nogrey as options, neither of which will be on by default. I've edited the commit message and pr to reflect this.

File: wadsrc/static/menudef.txt [hunk 1]
```diff
@@ -2617,6 +2617,14 @@ OptionMenu "PostProcessMenu" protected
 	StaticText " "
 	Slider "$GLPREFMNU_PALTONEMAPPOWER",		gl_paltonemap_powtable,			0.2, 3.0, 0.1, 1
 	Option "$GLPREFMNU_PALTONEMAPORDER",		gl_paltonemap_reverselookup,	"LookupOrder"
+	Option "$GLPREFMNU_PALTONECMAP",		gl_paltonemap_colormap,		"ColormapTypes"
+	Option "$GLPREFMNU_PALTONENOGREY",		gl_paltonemap_colormap_nogrey,	"YesNo"
+}
+
+OptionValue "ColormapTypes"
+{
+	0, "Neutral"
```
Comment (alexey-lysiuk) [inline wadsrc/static/menudef.txt]: A bit unrelated to feature itself, but this line requires localization.

Comment (axredneck) [reply wadsrc/static/menudef.txt]: "Нейтральная" in Russian

Comment (madame-rachelle) [reply wadsrc/static/menudef.txt]: > "Нейтральная" in Russian

You will have to put that on the spreadsheet when the strings get there. Putting it here is just going to lead to it getting lost - especially if other people do the same thing as you, here.

**Comment by Gaerzi** [comment]
> The following strings are needed for the menu
> 
> GLPREFMNU_PALTONECMAP = "Tonemap Palette Colormap";
> GLPREFMNU_PALTONENOGREY = "Prevent Grey Shadows";

What does "Tonemap Palette Colormap" mean exactly? It's not really clear.

**Comment by drfrag666** [comment]
> Extend UpdateTextures() so it can use the new function to generate a lut with a neutral colourmap
for the palette shader
The options are neutral and Doom which i presume was the original colormap.

**Comment by chetgurevitch** [comment]
@madame-rachelle the main thing is the neutral colormap, which is more accurate for games like heretic and hexen

**Comment by MajorCooke** [comment]
@madame-rachelle does this look alright for merging into QZDoom for testing? This one is a bit outside of my expertise...

**Comment by coelckers** [comment]
Since that bad text in here that turned out to be a blocker still hasn't been clarified, I am now closing the PR as a whole.

**Pull request closed without merging**